app/controllers/application_controller.rb

```diff
@@ -16,6 +16,19 @@ class ApplicationController < ActionController::Base
 
   protected
 
+  def current_admin_user
+    return current_user if current_user && current_user.role == "admin"
+    return false
+  end
+
+  def authenticate_admin_user!
+    if current_admin_user
+      flash[:notice] = "Hi #{current_user.name}"
+    else
+      redirect_to errors_path, notice: 'Admin role required'
```

Comment:
There is also the case where the user might not be logged in. You can't redirect him to the errors_path in that case. I know it will actually get him to login if that doesn't work but this is not good logic.